src/objc/AudioPreprocessing.m

```diff
@@ -0,0 +1,365 @@
+    _spectrum.realp = (float*)calloc(fftLength, sizeof(float));
+    _spectrum.imagp = (float*)calloc(fftLength, sizeof(float));
+    _PowerSpectrum = (float*)calloc(fftLength/2 + 1, sizeof(float));
+    _workingFrame = (float*)calloc(fftLength, sizeof(float));
+    
+    return self;
+}
+
+-(void)dealloc {
+    free(_hammingWindow);
+    vDSP_destroy_fftsetup(_fftSetup);
+    free(_spectrum.realp);
+    free(_spectrum.imagp);
+    free(_PowerSpectrum);
+    free(_workingFrame);
+}
+
+-(void)applyHammingWindow:(float*)signal numSamples:(NSUInteger)numSamples {
+    NSAssert(numSamples == _frameLength, @"Num samples and frame length must be the same.");
+    vDSP_vmul(_hammingWindow, 1, signal, 1, signal, 1, numSamples);
+}
+
+-(void)calculatePowerSpectrum:(float*)signal numSamples:(NSUInteger)numSamples {
+    // Copy signal into working frame.
+    memcpy(_workingFrame, signal, numSamples * sizeof(float));
+
+    // Pad signal with 0 to fftLength.
+    memset(&_workingFrame[numSamples], 0, (_fftLength - numSamples) * sizeof(float));
+
+    // Produce packed input necessary for vDSP_fft to work.
+    vDSP_ctoz((DSPComplex *)_workingFrame, 2, &_spectrum, 1, _fftLength/2);
+
+    // Perform the forward FFT.
+    vDSP_fft_zrip(_fftSetup, &_spectrum, 1, log2(_fftLength), kFFTDirection_Forward);
+
+    float scale = 0.5; // Factor of 2 needed due to vDSP_fft_zrip implementation detail.
+    vDSP_zvabs(&_spectrum, 1, _PowerSpectrum, 1, _fftLength/2);
+    vDSP_vsmul(_PowerSpectrum, 1, &scale, _PowerSpectrum, 1, _fftLength/2);
+}
+
+-(void)_generateHammingWindow:(NSUInteger)frameLength {
+    // Don't use vDSP_hamm_window because it uses a slightly different implementation
+    for(size_t i = 0; i < frameLength; i++) {
+        _hammingWindow[i] = 0.5 - 0.5*cos((2*M_PI*i)/(frameLength));
+    }
+}
+
+@end // Power Spectrum implementation
+
+
+@interface MelFrequencyFilterBank : NSObject
+
+@property (readonly) float* MelSpectrum;
+
+-(id)initWithMinFrequency:(float)minFreq
+             maxFrequency:(float)maxFrequency
+               sampleRate:(double)sampleRate
+              numMelBands:(NSUInteger)numMelBands
+                  numBins:(NSUInteger)numBins;
+
+-(void)apply:(const float*)powerSpectrum
+      output:(float *)out;
+
+@end   // MelFrequencyFilterBank interface
+
+
+@implementation MelFrequencyFilterBank {
+    float _minFreq, _maxFreq, _sampleRate;
+    NSUInteger _numMelBands, _numBins;
+    NSMutableArray* _filters;
+    float** _filterBanks;
+}
+
+-(id)initWithMinFrequency:(float)minFreq
+             maxFrequency:(float)maxFreq
+               sampleRate:(double)sampleRate
+              numMelBands:(NSUInteger)numMelBands
+                  numBins:(NSUInteger)numBins {
+
+    self = [super init];
+
+    _minFreq = minFreq;
+    _maxFreq = maxFreq;
+    _sampleRate = sampleRate;
+    _numMelBands = numMelBands;
+    _numBins = numBins;
+    _filterBanks = calloc(_numMelBands, sizeof(float *));
```

Comment:
Where is this deallocated?

Reply:
Note that you might consider performing allocations using `NSData` (https://developer.apple.com/documentation/foundation/nsdata?language=objc) so that it can handle memory management for you. It looks like _filterBanks could be `NSMutableArray<NSData*> *`

Reply:
Good catch. I've added a `dealloc`. Please take a look at it.

I actually think I probably just shouldn't be dynamically allocating any memory in this file (i.e. no calloc/malloc). I **think** I could just do everything with variable sized arrays. What do you think?

Reply:
By variable sized array do you mean e.g.
```
int size = ...;
float my_array[size];
```
?

Note that this is not actually part of the C++ standard (AFAIK), but it just a popular extension that popular compilers have implemented. Since this code will only be compiled on Apple platforms, it may be okay if it compiles now given our current toolchain, but I think you may as well stick with heap allocations, given that you need one anyway to store _filterBanks from initializer to calls to `apply`. 

FWIW, I think it's fairly standard to have e.g.
```
NSData *lowerBankData = [[NSData alloc] initWithLength:spectrumSize*sizeof(float)];
float* lowerBank = (float*)lowerBankData.mutableBytes;
```
Then Objective C will take care of memory management for you, acknowledging that the `lowerBank` pointer will only remain valid as long as `lowerBankData` is around.

Reply:
Yes, I believe we're talking about the same thing. Basically the value of `size` isn't know until runtime. From what I can tell, it's been a part of C since C99 but only a part of C++ since C++11.

Good point about this approach not working because of needing to store the values between the initializer and the call to `apply`. Clearly variable sized arrays are not going to work here.